src/renderers/shared/fiber/ReactFiberStack.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright 2013-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule ReactFiberStack
+ * @flow
+ */
+
+'use strict';
+
+import type { Fiber } from 'ReactFiber';
+
+export type StackCursor<T> = {
+  current: T
+};
+
+const warning = require('warning');
+
+const valueStack : Array<any> = [];
+
+if (__DEV__) {
+  var fiberStack : Array<Fiber | null> = [];
+}
+
+let index = -1;
+
+exports.createCursor = function<T>(defaultValue : T) : StackCursor<T> {
+  return {
+    current: defaultValue,
+  };
+};
+
+exports.isEmpty = function() : boolean {
+  return index === -1;
+};
+
+exports.pop = function<T>(
+  cursor : StackCursor<T>,
+  fiber: Fiber,
+) : void {
+  if (index < 0) {
+    if (__DEV__) {
+      warning(false, 'Unexpected pop.');
+    }
+    return;
+  }
+
+  if (__DEV__) {
+    if (fiber !== fiberStack[index]) {
+      warning(false, 'Unexpected Fiber popped.');
+    }
+  }
+
+  cursor.current = index > 0
+    ? valueStack[index]
+    : (null : any);
+
+  valueStack[index] = null;
+
+  if (__DEV__) {
+    fiberStack[index] = null;
+  }
+
+  index--;
+};
+
+exports.push = function<T>(
+  cursor : StackCursor<T>,
+  value : any,
+  fiber: Fiber,
+) : void {
+  index++;
+
+  valueStack[index] = cursor.current;
+
+  if (__DEV__) {
+    fiberStack[index] = fiber;
+  }
+
+  cursor.current = value;
+};
+
+exports.reset = function<T>(
+  cursor : StackCursor<T>,
```

Comment:
Passing a cursor here seems like an odd API since it's invalid to only reset one cursor. You have to reset all at once. It also causes the unsound assignment below.

Currently this only works because index is already at `-1` at the second call to reset so the loop is unnecessary.

I'd recommend just dropping the `cursor` argument and reset them separately if you think it's necessary to reset them.